File: src/types/problem.ts
import { CreateImageRequest } from "./image";
import { InputParameterResponse } from "./input_parameter";
import { OutputParameterResponse } from "./output_parameter";
import { OutputValueResponse } from "./output_value";

export type ProblemResponse = {
    id: number;
    prev_problem_id: number | null;
    is_final: boolean;
    name: string;
    description: string;
    image_id: number | null;
    created_at: string;
    updated_at: string;
}

export type ProblemFullResponse = {
    id: number;
    prev_problem_id: number | null;
    is_final: boolean;
    name: string;
    description: string;
    image_id: number | null;
    created_at: string;
    updated_at: string;
    input_parameters: Array<InputParameterResponse>;
    output_parameters: Array<OutputParameterResponse>;
    output_values: Array<OutputValueResponse>;
}

export type CreateProblemRequest = {
    prev_problem_id: number | null;
    is_final: boolean;
    name: string;
    description: string;
    image: CreateImageRequest | null;
}

export type CreateProblemResponse = {
    id: number;
    image_id: number | null;
}

export type UpdateProblemRequest = {
    name: string;
    description: string;
}

File: src-tauri/src/domain/entities/problem.rs
use serde::{Deserialize, Serialize};

use crate::domain::entities::{
    image::Image, input_parameter::InputParameter, output_parameter::OutputParameter,
    output_value::OutputValue,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Problem {
    pub id: i64,
    pub prev_problem_id: Option<i64>,
    pub is_final: bool,
    pub name: String,
    pub description: String,
    pub image_id: Option<i64>,
    pub created_at: String,
    pub updated_at: Option<String>,
    pub input_parameters: Vec<InputParameter>,
    pub output_parameters: Vec<OutputParameter>,
    pub output_values: Vec<OutputValue>,
    pub image: Option<Image>,
}
